### PR TITLE
Rest and Graphql upload client

### PR DESCRIPTION
**Pull request by erezmus**
Adds Rest and GraphQL client upload with tests that use mocked storage service according to api definitions.

### CODE REVIEW

File: client/src/FileUploadRecord.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace Libero\SharedServicesExperiment\Client;
+
+class FileUploadRecord
+{
+
+  private $link;
+  private $last_modified;
```
Comment (giorgiosironi) [inline client/src/FileUploadRecord.php]: `camelCase` is the convention for variables/fields usually?

File: client/src/GraphQLFileUploader.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            throw new InvalidArgumentException('File not found: ' . $sourcePath);
+        }
+
+        $finfo         = finfo_open(FILEINFO_MIME_TYPE);
+        $mimeType      = finfo_file($finfo, $sourcePath);
+        $contentLength = filesize($sourcePath);
+        $body          = stream_for(fopen($sourcePath, 'r'));
+        list(, $path)  = explode('/', $uploadPath, 2) + [1 => null];
+
+        $query = <<<'QUERY'
```
Comment (giorgiosironi) [inline client/src/GraphQLFileUploader.php:39]: is this supposed a schema that is supposed to be hardcoded in the client?

File: client/src/RestFileUploader.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+     * @throws InvalidArgumentException
+     * @throws FileUploaderException
+     */
+    public function uploadFile(string $sourcePath, string $uploadPath): FileUploadRecord
+    {
+        if (! file_exists($sourcePath)) {
+            throw new InvalidArgumentException('File not found: ' . $sourcePath);
+        }
+
+        $finfo         = finfo_open(FILEINFO_MIME_TYPE);
```
Comment (giorgiosironi) [inline client/src/RestFileUploader.php:47]: a library like `symfony/filesystem` would probably be more realistic, albeit with more overhead?

Comment (erezmus) [reply client/src/RestFileUploader.php]: If needed, can be added further down the line. Not sure what the advantage would be in this case

File: client/src/GraphQLFileUploader.php (hunk 2 above)
Comment (erezmus) [reply client/src/GraphQLFileUploader.php:39]: it defines the which mutations to use on the server side. So here 'UploadFile' uses the 'uploadFile' mutation (which is resolved in the storage service)

File: client/src/RestFileUploader.php (hunk 3 above)
Comment (erezmus) [reply client/src/RestFileUploader.php:47]: Possibly, but not sure whats he advantage at this stage.

**Review by diversemix — APPROVED** [review]
Let's get this in - because there are bound to be changes later :smile:

**Pull request merged**